core/templates/domain/statistics/learner-answer-details-backend-api.service.ts

```diff
@@ -16,52 +17,59 @@
+import { Injectable } from '@angular/core';
+import { HttpClient } from '@angular/common/http';
+import { downgradeInjectable } from '@angular/upgrade/static';
 
-require('domain/statistics/statistics-domain.constants.ajs.ts');
+import { UrlInterpolationService } from
+  'domain/utilities/url-interpolation.service';
+import { StatisticsDomainConstants } from
+  'domain/statistics/statistics-domain.constants';
 
-angular.module('oppia').factory('LearnerAnswerDetailsBackendApiService', [
-  '$http', '$q', 'UrlInterpolationService', 'SUBMIT_LEARNER_ANSWER_DETAILS_URL',
-  function(
-      $http, $q, UrlInterpolationService, SUBMIT_LEARNER_ANSWER_DETAILS_URL) {
-    var _recordLearnerAnswerDetails = function(
-        expId, stateName, interactionId, answer, answerDetails,
-        successCallback, errorCallback) {
-      var recordLearnerAnswerDetailsUrl = (
-        UrlInterpolationService.interpolateUrl(
-          SUBMIT_LEARNER_ANSWER_DETAILS_URL, {
-            entity_type: 'exploration',
-            entity_id: expId
-          }));
+@Injectable({
+  providedIn: 'root'
+})
+export class LearnerAnswerDetailsBackendApiService {
+  constructor(
+        private http: HttpClient,
+        private urlInterpolationService: UrlInterpolationService
+  ) { }
 
-      var payload = {
-        state_name: stateName,
-        interaction_id: interactionId,
-        answer: answer,
-        answer_details: answerDetails
-      };
+  private _recordLearnerAnswerDetails(
+      expId: string, stateName: string, interactionId: string,
+      answer: string, answerDetails: string,
+      successCallback: any, errorCallback: any) {
+    let recordLearnerAnswerDetailsUrl = this.urlInterpolationService
+      .interpolateUrl(
+        StatisticsDomainConstants.SUBMIT_LEARNER_ANSWER_DETAILS_URL, {
+          entity_type: 'exploration',
+          entity_id: expId
+        });
 
-      $http.put(recordLearnerAnswerDetailsUrl, payload).then(function(
-          response) {
-        if (successCallback) {
-          successCallback();
-        }
-      }, function(errorResponse) {
-        if (errorCallback) {
-          errorCallback(errorResponse.data);
-        }
-      });
+    let payload = {
+      state_name: stateName,
+      interaction_id: interactionId,
+      answer: answer,
+      answer_details: answerDetails
     };
 
-    return {
-      recordLearnerAnswerDetails: function(expId, stateName, interactionId,
-          answer, answerDetails) {
-        return $q(function(resolve, reject) {
-          _recordLearnerAnswerDetails(
-            expId, stateName, interactionId, answer, answerDetails,
-            resolve, reject);
-        });
-      }
-    };
+    this.http.put(recordLearnerAnswerDetailsUrl, payload)
+      .toPromise().then(
+        () => successCallback && successCallback(),
+        errorResponse => errorCallback && errorCallback(errorResponse.body)
+      );
   }
-]);
+
+  recordLearnerAnswerDetails(
```

Comment:
Just move the code from `_recordLearnerAnswerDetails` into this function, and do not take any callbacks as arguments.

Reply:
Hey @brianrodri, Do you have any idea why was it added this way originally (in AngularJS) and shouldn't we migrate with the same structure as it is? 

For all migrated services we have kept the same structure with functions starting with `_` marked as private. Thoughts?

Reply:
Nope I've never understood why, it hasn't ever made much sense to me unfortunately. All I could do is guess why, and I know that starting functions with `_` was used to fake private methods, which is why I'd suggest removing them.

Reply:
Yeah, even I couldn't get why fake private functions which were a copy of public functions added. Do we want to remove them now? 

My major concern is I've made many such PR's to which new contributors are referring and in all those previous PR's we have kept both the functions. So all PR's (at least for sometime) will be issued with both these functions and then we will ask to revert it to one function. Wouldn't that be a bit messy?

Reply:
Hmm but we can just create a new issue to track and "collapse" them, so I don't think it'll be messy as long as we make it possible to fix (which sounds easy since this pattern seems grep-able). WDYT?

Reply:
To be honest, I'm not sure. We have migrated about more than 100 services(so we have to iterate over that) and removing such functions could be a bit confusing for folks, IMO. Also, we yet don't know the reason as to why these similar functions were added.

cc/ @seanlip @ankita240796

Reply:
@brianrodri is in charge of the learner analytics files. I defer to him for guidance on refactoring this part specifically.

I feel like that refactor shouldn't be part of the migration though?

I also don't think there's a need for a general issue to be filed. I think sometimes this pattern has its uses if other internal functions want to call the internal method too.

Reply:
Yeah, I'm along the lines that refactor shouldn't be part of migration.

Reply:
Yeah I don't think we should have it as part of the migration, but I'd prefer we remove them so I'd prefer to keep asking for that cleanup (as far as my codeowner files are concerned)

Reply:
I do not advocate a general cleanup of this pattern across the codebase, for reasons mentioned above. I'm fine for this to be done for specific files if @brianrodri wants (but perhaps @brianrodri you might want to file the issue in this case since you'll be defining what you want this file to be).